controllers/imageclusterinstall_controller.go

```diff
@@ -216,15 +225,126 @@ func (r *ImageClusterInstallReconciler) Reconcile(ctx context.Context, req ctrl.
 			return ctrl.Result{}, err
 		}
 
-		if err := r.setClusterInstalledConditions(ctx, ici); err != nil {
-			log.WithError(err).Error("failed to set installed conditions")
+		timedout, err := r.checkClusterTimeout(ctx, log, ici, r.DefaultInstallTimeout)
+		if err != nil {
+			log.WithError(err).Error("failed to check for install timeout")
+			return ctrl.Result{}, err
+		}
+		if timedout {
+			return ctrl.Result{}, nil
+		}
+
+		res, err := r.checkClusterStatus(ctx, log, ici, clusterDeployment)
+		if err != nil {
+			log.WithError(err).Error("failed to check cluster status")
 			return ctrl.Result{}, err
 		}
+		if !res.IsZero() {
+			return res, nil
+		}
+	}
+
+	return ctrl.Result{}, nil
+}
+
+func (r *ImageClusterInstallReconciler) checkClusterTimeout(ctx context.Context, log logrus.FieldLogger, ici *v1alpha1.ImageClusterInstall, defaultTimeout time.Duration) (bool, error) {
+	timeout := defaultTimeout
+
+	if timeoutOverride, present := ici.Annotations[installTimeoutAnnotation]; present {
+		var err error
+		timeout, err = time.ParseDuration(timeoutOverride)
+		if err != nil {
+			return false, fmt.Errorf("failed to parse install timeout annotation value %s: %w", timeoutOverride, err)
+		}
+	}
+
+	if ici.Status.BootTime.Add(timeout).Before(time.Now()) {
+		log.Error("timed out waiting for cluster to finish installation")
+		err := r.setClusterTimeoutConditions(ctx, ici, timeout.String())
+		if err != nil {
+			log.WithError(err).Error("failed to set cluster timeout conditions")
+		}
+		return true, err
+	}
+
+	return false, nil
+}
+
+func (r *ImageClusterInstallReconciler) checkClusterStatus(ctx context.Context, log logrus.FieldLogger, ici *v1alpha1.ImageClusterInstall, clusterDeployment *hivev1.ClusterDeployment) (ctrl.Result, error) {
+	spokeClient, err := r.spokeClient(ctx, ici)
+	if err != nil {
+		log.WithError(err).Error("failed to create spoke client")
+		return ctrl.Result{}, err
+	}
+	installed, err := r.MonitorInstall(ctx, log, spokeClient)
```

Comment:
Cnosider renaming to:
 ```suggestion
	installed, err := r.GetSpokeClusterInstallStatus(ctx, log, spokeClient)
```
You can probably move the spoke client initialization into that function as well (it's the only one using it)

Reply:
I'm fine with the rename, but I'd rather not move the spoke client stuff because this makes it a lot easier to mock the spoke client in the monitor code. Otherwise I need to create some kind of factory so I can mock it to return the fake client.